Source/Dafny/AST/DafnyAst.cs

```diff
@@ -9070,6 +9086,48 @@ public abstract class Expression {
       get { yield break; }
     }
 
+    private IToken startTok;
+    private IToken endTok;
+
+    // Creates a token on the entire range of the expression.
+    // Used only for error reporting.
+    public IToken RangeToken => new RangeToken(StartTok, EndTok);
+
+    // Computes the start and end token of this expression
+    private void RecomputeStartEnd() {
+      startTok = tok;
+      endTok = tok;
+      foreach (var e in SubExpressions) {
```

Comment:
Possibly out of scope: it would help if `SubExpressions` was a `IReadOnlyList` so you could retrieve the first and last element without having to iterate.

Reply:
You mean, we could cache the SubExpressions? That's a good point, but yes, let's leave it out of scope for this PR.